### DISCUSSION TITLE
Creating a user preferences table

### DISCUSSION THREAD

**Discussion by techthrasher** [Budibase/budibase#12825] [Help] [answered]
I've looked but can't find anything particular to this so please forgive me if it's so basic I should already know it.

I'd like the app user to be able set a few user-global preferences that I can then use to present the data in ways most useful to them.  I would have like to just add a column to the users tables and accessed it that way, but that functionality was removed (for what I can only imagine were very good and very technical reasons).  

So I'm left with creating a user preferences table, and tying the data to the user with a "created-by" column.  I could probably implement at some basic level, but I'm concerned it will create 1) performance problems (additional data provider for every. single. form.) and also potentially difficult to implement in more advanced ways (referencing a user preference to execute a Javascript expression).

The questions:  Would introducing a user preferences table likely introduce performance problems?  Can the data in the user preferences table be exposed to essentially all the functionality in Budibase and javascript?

Thank you so much for taking the time to answer my numpty questions....

**Comment by ConorWebb96** [comment]
Hey @techthrasher,

I'd recommend you have a look at the [user column](https://docs.budibase.com/docs/users-1). This might be exactly what you need.

**Reply by techthrasher** [reply]
Thanks @ConorWebb96 !

Yep, I've implemented the user column and I'm glad to have it.  This is perhaps more of a design question than functionality - I know I CAN create a user preferences table, the question is SHOULD I create a user preferences table?  Followed up by, if I DO create a user preferences table, will I be able to implement it effectively in custom scripting? 

Here's the example: you helped me a couple of weeks ago on putting together a script that converts hours stored in the database to days, hours and minutes.  That was incredibly helpful.  But the JS assumes a day is 24 hours for all users.  I'd like the user to be able to specify they would prefer to use 4, 8 or 12 hours as a "day" (it's about task management).  So I can create a user preference table where the user says, "hey, I only want you to allocate 8 hours as a 'day' for my task management", but then will the JS be able to use that variable in the script?  If so, will that JS need to have a data provider for that variable or can it just look it up as part of the script (to avoid having too many data providers on a screen)?

I'm sorry if I'm making a mess of this.  It doesn't help that I'm only beginning with JS and Budibase (SQL I've spend some time with...)

**Reply by ConorWebb96** [reply, marked as answer]
Hey @techthrasher,

I can't see a bad reason for creating a table like you've described. It really depends how many different options will be provided in the table. I can't see a few 100 options negatively effecting performance. You can easily target the current user through `Current User` bindings. This means that you can get the related data for that specific user. Meaning only their records would be pulled in.
<img width="311" alt="Screenshot 2024-01-23 at 10 07 41" src="https://github.com/Budibase/budibase/assets/126772285/4a5c6c3f-f816-4636-ad3e-5cf99210f37c">

The JS should be able to use whatever data you provide it in your case it may look something like this; 
```js 
$("preferences table.day hours") // from the filter user preferences data provider
```
You can just add this in place of 24 and it will use whatever selected value is stored within instead. (Obviously you may want to do some validation on the binding so that strings can't be passed in etc.)

If you are worried about performance you can always use actions in place of data providers in some cases. For example you could use a [fetch row](https://docs.budibase.com/docs/data-actions#fetch-row) action with an update state action to only get specific rows/fragments of information as and when you need it. This can be run off on screen load or even button clicks.

[Execute query](https://docs.budibase.com/docs/data-actions#execute-query) is another option, this can be tied in with the [public api](https://docs.budibase.com/reference/rowsearch) to get really specific collections of data. As it uses the rest connector you can use the in-built transformer to really refine the data you receive back which can then be used in combination with state throughout the page.

I hope this helps to answer your questions somewhat.

**Reply by techthrasher** [reply]
@ConorWebb96 - thank you so much for the steer - definitely helps.  I really appreciate there are some dedicated and talented folks going through the discussion board to help.  

I'm not terribly worried about performance, but I've seen many references in the documentation/discussion to "don't do this because it may negatively impact performance."  So I'm a bit skittish to color outside the lines without permission....
